### ISSUE TITLE
Index Payments by Sequence Number

### ISSUE THREAD

**Issue by carlaKC** [lightningnetwork/lnd#4164]
Our current payments database is keyed by payment hash. This provides a series of challenges when trying to paginate queries on the data:
- Keys are stored in lexicographical order in bbolt, so insertion of new payments alters the order in which they are stored in the main bucket 
- We have a `includeIncomplete` parameter (false by default for backwards compatibility) which requires that we read each payment from disk to determine whether it should be part of our result set

#### As Is
ListPayments queries all payments from the database then filters to return a paginated response to the user. This was required because some nodes were hitting the max message size for grpc messages, and were unable to use the `ListPayments` endpoint at all. 

#### Steps to completion:
- Create a new bucket `payments-index`, keyed by sequence number to hold an index of the existing payments bucket. 
```
[payments-index]
    sequence number: payment hash
    sequence number: payment hash
```
- Add a migration which populates this database for all existing payments (fetch payments, write to index) 
- Update `InitPayment` to populate the `payments-index` table when a new payment in initiated 
- Update current listpayments to use this index rather than fetching all payments from disk then filtering them 
    ```
    // Start with cursor provided in query
    cursor := start sequence nr provided in query

    // While we are under our max nr of payments, and we have 
    // entries remaining in the index bucket
    while nr Payments < max; cursor !=nil
      // Get the payment hash for this sequence nr.
      pmtHash:= indexBucket.Get(cursor) 

      // Get the payment keyed by this hash from the
      // payments bucket.
      payment := fetchPayment(pmtHash)

     // Existing logic checking incomplete etc
    ```

**Comment by joostjager** [comment]
I think it may be worth to, before moving on with further code changes, to write down the ideal proto for both payments and invoices. So far everything has grown mostly in an organic way, which isn't always ideal.

Payment and invoices are very similar. Both are objects that go through a life-cycle. They can be listed (`ListPayments`, `ListInvoices`, `LookupInvoice`) and tracked (`TrackPayment`, `SubscribeInvoices`, `SubscribeSingleInvoice`). It should be possible to come up with one design that works for both of them.

That design doesn't need to be implemented right away, but it can help in defining smaller steps like what is proposed in this issue.

**Comment by joostjager** [comment]
Additional discussion on this "design" pr: https://github.com/lightningnetwork/lnd/pull/4166

**Comment by Roasbeef** [comment]
@joostjager with what was outlined in this issue, the RPC interface wouldn't change at all, so this ideal proto discussion is completely distinct. Also there's likely a significant difference between the ideal protos from your perspective (a developer of lnd) vs the perspective of the major users that are actually _consuming_ this API.

**Comment by joostjager** [comment]
Yes, RPC interface won't change at all with what is outlined in this issue. But if the RPC interface as it is now doesn't support certain use cases, I think it can be worth identifying the deficits and knowingly accept them or plan to make changes to the RPC. In the latter case, those changes may invalidate this issue. So I don't agree that the ideal proto discussion is completely distinct.

>Also there's likely a significant difference between the ideal protos from your perspective (a developer of lnd) vs the perspective of the major users that are actually consuming this API.

That can always be the case. Therefore I think it is good to sometimes look at the big picture instead of only the nearest improvement. And to share ideas like "ideal protos" with major users for feedback. It is a very limited time investment compared to coding, test and supporting anything in an implementation.

**Comment by joostjager** [comment]
Ideal proto discussion hasn't yet resulted in the ideal solution. One insight that developed is that `ListPayments` could evolve in the future to (in addition to reporting the current state) stream back live updates for all payments within the given index range. A monitoring or wallet app could use this to show the live status of a page of payments without polling.

The steps outlined in the issue description are definitely a logical step to tame memory usage. If the problem is pressing, we should go ahead with adding the index.

**Comment by cfromknecht** [comment]
Can this alternatively be addressed by storing the payment hash in the invoice body?

**Comment by cfromknecht** [comment]
Closed as of #4261
